### PR TITLE
Adding Example-4: Adding robot with individual sensor

### PR DESCRIPTION
**Pull request by destogl**

### CODE REVIEW

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#include "hardware_interface/types/hardware_interface_return_values.hpp"
+#include "hardware_interface/types/hardware_interface_status_values.hpp"
+#include "rclcpp/macros.hpp"
+#include "ros2_control_demo_hardware/visibility_control.h"
+
+using hardware_interface::return_type;
+
+namespace ros2_control_demo_hardware
+{
+class RRBotSystemWithSensorHardware
```
Comment (destogl) [inline ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp:39]: We don't need a new robot implementation, but only implementation of a sensor.

**Review by destogl — COMMENTED** [review]
Check the individual comments.

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+: public hardware_interface::BaseInterface<hardware_interface::SystemInterface>
```
Comment (destogl) [inline ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp:40]:
```suggestion
: public hardware_interface::BaseInterface<hardware_interface::SensorInterface>
```

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  return_type start() override;
+
+  ROS2_CONTROL_DEMO_HARDWARE_PUBLIC
+  return_type stop() override;
+
+  ROS2_CONTROL_DEMO_HARDWARE_PUBLIC
+  return_type read() override;
+
+  ROS2_CONTROL_DEMO_HARDWARE_PUBLIC
+  return_type write() override;
```
Comment (destogl) [inline ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp:64]:
```suggestion
```

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  RCLCPP_SHARED_PTR_DEFINITIONS(RRBotSystemWithSensorHardware);
+
+  ROS2_CONTROL_DEMO_HARDWARE_PUBLIC
+  return_type configure(const hardware_interface::HardwareInfo & info) override;
+
+  ROS2_CONTROL_DEMO_HARDWARE_PUBLIC
+  std::vector<hardware_interface::StateInterface> export_state_interfaces() override;
+
+  ROS2_CONTROL_DEMO_HARDWARE_PUBLIC
+  std::vector<hardware_interface::CommandInterface> export_command_interfaces() override;
```
Comment (destogl) [inline ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp:52]:
```suggestion
```

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  return_type read() override;
+
+  ROS2_CONTROL_DEMO_HARDWARE_PUBLIC
+  return_type write() override;
+
+private:
+  // Parameters for the RRBot simulation
+  double hw_start_sec_;
+  double hw_stop_sec_;
+  double hw_slowdown_;
```
Comment (destogl) [inline ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp:70]:
```suggestion
```

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  double hw_sensor_change_;
+
+  // Store the command for the simulated robot
+  std::vector<double> hw_joint_commands_;
+  std::vector<double> hw_joint_states_;
```
Comment (destogl) [inline ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp:75]:
```suggestion
```

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  return_type write() override;
+
+private:
+  // Parameters for the RRBot simulation
+  double hw_start_sec_;
+  double hw_stop_sec_;
+  double hw_slowdown_;
+  double hw_sensor_change_;
+
+  // Store the command for the simulated robot
```
Comment (destogl) [inline ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp:73]:
```suggestion
```

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+private:
+  // Parameters for the RRBot simulation
+  double hw_start_sec_;
+  double hw_stop_sec_;
+  double hw_slowdown_;
+
+  // Store the command for the simulated robot
+  std::vector<double> hw_commands_;
+  std::vector<double> hw_states_;
+};
```
Comment (destogl) [inline ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp:115]:
```suggestion
};
```

File: ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+  const hardware_interface::HardwareInfo & info)
+{
+  if (configure_default(info) != hardware_interface::return_type::OK)
+  {
+    return hardware_interface::return_type::ERROR;
+  }
+
+  hw_start_sec_ = stod(info_.hardware_parameters["example_param_hw_start_duration_sec"]);
+  hw_stop_sec_ = stod(info_.hardware_parameters["example_param_hw_stop_duration_sec"]);
+  hw_slowdown_ = stod(info_.hardware_parameters["example_param_hw_slowdown"]);
```
Comment (destogl) [inline ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp:43]:
```suggestion
```

File: ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+#include <limits>
+#include <memory>
+#include <vector>
+
+#include "hardware_interface/types/hardware_interface_type_values.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+namespace ros2_control_demo_hardware
+{
+hardware_interface::return_type RRBotSystemPositionOnlyHardware::configure(
```
Comment (destogl) [inline ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp:33]:
```suggestion
hardware_interface::return_type DummyForceTorqueSensorHardware::configure(
```

File: ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+
+    if (joint.state_interfaces[0].name != hardware_interface::HW_IF_POSITION)
+    {
+      RCLCPP_FATAL(
+        rclcpp::get_logger("RRBotSystemPositionOnlyHardware"),
+        "Joint '%s' have %s state interface. '%s' expected.", joint.name.c_str(),
+        joint.state_interfaces[0].name.c_str(), hardware_interface::HW_IF_POSITION);
+      return hardware_interface::return_type::ERROR;
+    }
+  }
```
Comment (destogl) [inline ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp:85]: Add sensor part from the Example 3 (#112)

```suggestion
```

File: ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+  std::vector<hardware_interface::CommandInterface> command_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++)
+  {
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(
+      info_.joints[i].name, hardware_interface::HW_IF_POSITION, &hw_commands_[i]));
+  }
+
+  return command_interfaces;
+}
+
```
Comment (destogl) [inline ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp:116]:
```suggestion
```

File: ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+  std::vector<hardware_interface::CommandInterface> command_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++)
+  {
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(
+      info_.joints[i].name, hardware_interface::HW_IF_POSITION, &hw_joint_commands_[i]));
+  }
+
+  return command_interfaces;
+}
+
```
Comment (destogl) [inline ros2_control_demo_hardware/src/rrbot_system_with_external_sensor.cpp:314]:
```suggestion
```

File: ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro [hunk 14]
```diff
@@ -0,0 +1,82 @@
+<?xml version="1.0"?>
```
[the remark this answers is not in the document]
Comment (destogl) [reply ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro:1]: Actually, the file can be renamed to `dummy_force_torque_sensor.ros2_control.xacro`. And it holds only macro for the sensors.

File: ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
+
+  <xacro:macro name="rrbot_system_with_external_sensor" params="name prefix use_sim:=^|false use_fake_hardware:=^|true fake_sensor_commands:=^|false slowdown:=2.0">
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro:4]:
```suggestion
  <xacro:macro name="dummy_force_torque_sensor" params="name prefix use_sim:=^|false use_fake_hardware:=^|true fake_sensor_commands:=^|false">
```

File: ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+      </joint>
+      <joint name="joint2">
+        <command_interface name="position">
+          <param name="min">-1</param>
+          <param name="max">1</param>
+        </command_interface>
+        <state_interface name="position"/>
+      </joint>
+    </ros2_control>
+
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro:44]:
```suggestion
```

File: ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    name="RRBotSystemPositionOnly" prefix="$(arg prefix)"
+    use_fake_hardware="$(arg use_fake_hardware)"
+    fake_sensor_commands="$(arg fake_sensor_commands)"
+    slowdown="$(arg slowdown)" />
+
+  <xacro:rrbot_system_with_external_sensor
+    name="RRBotSystemWithSensor" prefix="$(arg prefix)"
+    use_fake_hardware="$(arg use_fake_hardware)"
+    fake_sensor_commands="$(arg fake_sensor_commands)"
+    slowdown="$(arg slowdown)" />
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro:50]:
```suggestion
  <xacro:external_rrbot_force_torque_sensor
    name="ExternalRRBotFTSensor" prefix="$(arg prefix)"
    use_fake_hardware="$(arg use_fake_hardware)"
    fake_sensor_commands="$(arg fake_sensor_commands)" />
```

File: ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+  <xacro:include filename="$(find ros2_control_demo_description)/urdf/rrbot/rrbot.urdf.xacro" />
+
+  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
+  <xacro:include filename="$(find ros2_control_demo_description)/urdf/rrbot/gazebo/rrbot.gazebo.xacro" />
+
+  <!-- Import Rviz colors -->
+  <!-- <xacro:include filename="$(find ros2_control_demo_description)/urdf/rrbot/gazebo/rrbot.materials.xacro" /> -->
+
+  <!-- Import RRBot ros2_control description -->
+  <xacro:include filename="$(find ros2_control_demo_description)/urdf/rrbot/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro" />
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro:26]:
```suggestion
  <xacro:include filename="$(find rrbot_description)/ros2_control/rrbot_system.ros2_control.xacro" />
  <xacro:include filename="$(find rrbot_description)/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro" />
```

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp [hunk 19]
```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2021, Stogl Robotics Consulting UG (haftungsbeschränkt)
```
Comment (destogl) [inline ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_with_external_sensor.hpp:1]: Rename to `external_rrbot_force_torque_sensor.hpp`

File: ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro (hunk 14 above)
[the remark this answers is not in the document]
Comment (bailaC) [reply ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro:1]: Changed the file to `external_rrbot_force_torque_sensor.ros2_control.xacro` and kept only the sensor definition.

File: ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro (hunk 15 above)
Comment (bailaC) [reply ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro:4]: Done.

File: ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro (hunk 16 above)
Comment (bailaC) [reply ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro:44]: Done!

File: ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+          <xacro:if value="${use_fake_hardware}">
+            <plugin>fake_components/GenericSystem</plugin>
+            <param name="fake_sensor_commands">${fake_sensor_commands}</param>
+            <param name="state_following_offset">0.0</param>
+          </xacro:if>
+          <xacro:unless value="${use_fake_hardware}">
+            <plugin>ros2_control_demo_hardware/RRBotSystemWithSensorHardware</plugin>
+            <param name="example_param_hw_start_duration_sec">2.0</param>
+            <param name="example_param_hw_stop_duration_sec">3.0</param>
+            <param name="example_param_hw_slowdown">${slowdown}</param>
```
Comment (bailaC) [inline ros2_control_demo_description/rrbot_description/ros2_control/rrbot_system_with_external_sensor.ros2_control.xacro:62]: Should we remove the Line : 62 as well?

File: ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro [hunk 21]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0"?>
```
[the remark this answers is not in the document]
Comment (bailaC) [reply ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro:1]: Took the structure from Example 1 for now.

File: ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+          <xacro:if value="${use_fake_hardware}">
+            <plugin>fake_components/GenericSystem</plugin>
+            <param name="fake_sensor_commands">${fake_sensor_commands}</param>
+            <param name="state_following_offset">0.0</param>
+          </xacro:if>
+          <xacro:unless value="${use_fake_hardware}">
+            <plugin>ros2_control_demo_hardware/ExternalRRBotForceTorqueSensorHardware</plugin>
+            <param name="example_param_hw_start_duration_sec">2.0</param>
+            <param name="example_param_hw_stop_duration_sec">3.0</param>
+            <param name="example_param_hw_slowdown">${slowdown}</param>
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro:23]: This Parameter is not needed for sensor

File: ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+
+      <sensor name="tcp_fts_sensor">
+        <state_interface name="force.x"/>
+        <state_interface name="force.y"/>
+        <state_interface name="force.z"/>
+        <state_interface name="torque.x"/>
+        <state_interface name="torque.y"/>
+        <state_interface name="torque.z"/>
+        <param name="frame_id">tool_link</param>
+        <param name="lower_limits">100</param>
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro:37]: Please adjust limit-parameters to example 3. Names and values

File: ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+<!-- Revolute-Revolute Manipulator with a 2D Force-Torque sensor -->
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro:2]:
```suggestion
<!-- Revolute-Revolute Manipulator with a 6D Force-Torque sensor -->
```

File: ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+
+  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
+  <xacro:include filename="$(find rrbot_description)/gazebo/rrbot.gazebo.xacro" />
+
+  <!-- Import Rviz colors -->
+  <xacro:include filename="$(find rrbot_description)/gazebo/rrbot.materials.xacro" />
+
+  <!-- Import RRBot ros2_control description -->
+  <xacro:include filename="$(find rrbot_description)/ros2_control/rrbot_system.ros2_control.xacro" />
+
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro:27]: Here is missing include of sensor xacro file

File: ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+
+//
+// Authors: Subhas Das, Denis Stogl
+//
+
+#include "ros2_control_demo_hardware/external_rrbot_force_torque_sensor.hpp"
+
+#include <chrono>
+#include <cmath>
+#include <iostream>
```
Comment (destogl) [inline ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp:23]:
```suggestion
```
 Should not be here

File: ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+hardware_interface::return_type ExternalRRBotForceTorqueSensorHardware::configure(
+  const hardware_interface::HardwareInfo & info)
+{
+  if (configure_default(info) != hardware_interface::return_type::OK)
+  {
+    return hardware_interface::return_type::ERROR;
+  }
+
+  hw_start_sec_ = stod(info_.hardware_parameters["example_param_hw_start_duration_sec"]);
+  hw_stop_sec_ = stod(info_.hardware_parameters["example_param_hw_stop_duration_sec"]);
```
Comment (destogl) [inline ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp:42]: Missing ‘sensor change’ parameter

File: ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+  RCLCPP_INFO(
+    rclcpp::get_logger("ExternalRRBotForceTorqueSensorHardware"), "System successfully stopped!");
+
+  return hardware_interface::return_type::OK;
+}
+
+hardware_interface::return_type ExternalRRBotForceTorqueSensorHardware::read()
+{
+  RCLCPP_INFO(rclcpp::get_logger("ExternalRRBotForceTorqueSensorHardware"), "Reading...");
+
```
Comment (destogl) [inline ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp:91]: You should here calculate sensor values to return. Why are they not set at all?

File: ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro (hunk 22 above)
Comment (bailaC) [reply ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro:23]: `example_param_hw_slowdown `parameter has been removed.

File: ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro (hunk 23 above)
Comment (bailaC) [reply ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro:37]: Corrected.

File: ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro (hunk 25 above)
Comment (bailaC) [reply ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro:27]: added.

File: ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp (hunk 27 above)
Comment (bailaC) [reply ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp:41]: Added.

File: ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp (hunk 28 above)
Comment (bailaC) [reply ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp:90]: I removed it by mistake. Added now.

File: ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+  return hardware_interface::return_type::OK;
+}
+
+hardware_interface::return_type ExternalRRBotForceTorqueSensorHardware::stop()
+{
+  RCLCPP_INFO(
+    rclcpp::get_logger("ExternalRRBotForceTorqueSensorHardware"),
+    "Stopping ...please wait...");
+
+  for (int i = 0; i <= hw_stop_sec_; i++) {
```
Comment (destogl) [inline ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp:74]:
```suggestion
  for (int i = 0; i < hw_stop_sec_; i++) {
```

File: ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+  return hardware_interface::return_type::OK;
+}
+
+hardware_interface::return_type ExternalRRBotForceTorqueSensorHardware::start()
+{
+  RCLCPP_INFO(
+    rclcpp::get_logger("ExternalRRBotForceTorqueSensorHardware"),
+    "Starting ...please wait...");
+
+  for (int i = 0; i <= hw_start_sec_; i++) {
```
Comment (destogl) [inline ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp:53]:
```suggestion
  for (int i = 0; i < hw_start_sec_; i++) {
```

File: ros2_control_demo_bringup/launch/rrbot_system_with_external_sensor.launch.py [hunk 31]
```diff
@@ -0,0 +1,89 @@
...
+    # collect all the nodes here
+    nodes = [
+        fts_broadcaster_spawner,
+    ]
+
+    base_launch = IncludeLaunchDescription(
+        PythonLaunchDescriptionSource([ThisLaunchFileDir(), "/rrbot.launch.py"]),
+        launch_arguments={
+            "controllers_file": "rrbot_with_sensor_controllers.yaml",
+            "description_file": "rrbot_system_with_sensor.urdf.xacro",
```
[the remark this answers is not in the document]
Comment (destogl) [reply ros2_control_demo_bringup/launch/rrbot_system_with_external_sensor.launch.py:81]:
```suggestion
            "description_file": "rrbot_system_with_external_sensor.urdf.xacro",
```

File: ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro [hunk 32]
```diff
@@ -0,0 +1,39 @@
...
+    <ros2_control name="${name}" type="sensor">
+      <hardware>
+        <xacro:if value="${use_fake_hardware}">
+          <plugin>fake_components/GenericSystem</plugin>
+          <param name="fake_sensor_commands">${fake_sensor_commands}</param>
+          <param name="state_following_offset">0.0</param>
+        </xacro:if>
+        <xacro:unless value="${use_fake_hardware}">
+          <plugin>ros2_control_demo_hardware/ExternalRRBotForceTorqueSensorHardware</plugin>
+          <param name="example_param_hw_start_duration_sec">2.0</param>
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro:15]: Avoid RViz warning on start.

```suggestion
          <param name="example_param_hw_start_duration_sec">0.0</param>
```

File: ros2_control_demo_bringup/config/rrbot_with_external_sensor_controllers.yaml [hunk 33]
```diff
@@ -0,0 +1,25 @@
...
+  ros__parameters:
+    joints:
+      - joint1
+      - joint2
+    interface_name: position
+
+fts_broadcaster:
+  ros__parameters:
+    interface_names.force.x: tcp_fts_sensor/force.x
+    interface_names.torque.z: tcp_fts_sensor/torque.z
```
Comment (destogl) [inline ros2_control_demo_bringup/config/rrbot_with_external_sensor_controllers.yaml:24]: Lets user here 6D sensor with definition of sensor name using then Semantic Component.

```suggestion
    sensor_name: tcp_fts_sensor
```

File: ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro [hunk 34]
```diff
@@ -0,0 +1,48 @@
...
+https://github.com/ros-simulation/gazebo_ros_demos/blob/kinetic-devel/rrbot_description/urdf/rrbot.xacro
+-->
+<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="2dof_robot">
+  <xacro:arg name="use_sim" default="false" />
+
+  <!-- Enable setting arguments from the launch file -->
+  <xacro:arg name="use_fake_hardware" default="true" />
+  <xacro:arg name="fake_sensor_commands" default="false" />
+  <xacro:arg name="prefix" default="" />
+  <xacro:arg name="slowdown" default="2.0" />
```
Comment (destogl) [inline ros2_control_demo_description/rrbot_description/urdf/rrbot_system_with_external_sensor.urdf.xacro:14]: As of #126 for nicer behavior

```suggestion
  <xacro:arg name="slowdown" default="50.0" />
```

**Review by destogl — COMMENTED** [review]
Please apply other comments as in #112

File: ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp [hunk 35]
```diff
@@ -0,0 +1,131 @@
...
+{
+  if (configure_default(info) != hardware_interface::return_type::OK)
+  {
+    return hardware_interface::return_type::ERROR;
+  }
+
+  hw_start_sec_ = stod(info_.hardware_parameters["example_param_hw_start_duration_sec"]);
+  hw_stop_sec_ = stod(info_.hardware_parameters["example_param_hw_stop_duration_sec"]);
+  hw_sensor_change_ = stod(info_.hardware_parameters["example_param_max_sensor_change"]);
+
```
Comment (destogl) [inline ros2_control_demo_hardware/src/external_rrbot_force_torque_sensor.cpp:43]: @bailaC This should solve the crashing issue.

```suggestion

  hw_sensor_states_.resize(
    info_.sensors[0].state_interfaces.size(), std::numeric_limits<double>::quiet_NaN());

```

File: README.md [hunk 36]
```diff
@@ -329,6 +329,33 @@ angular:
...
+- ros2_control URDF: [external_rrbot_force_torque_sensor.ros2_control.xacro](ros2_control_demo_description/rrbot_description/ros2_control/external_rrbot_force_torque_sensor.ros2_control.xacro)
+
+- Command interfaces:
+  - joint1/position
+  - joint2/position
+- State interfaces:
+  - joint1/position
+  - joint2/position
+  - tcp_fts_sensor/force.x
+  - tcp_fts_sensor/torque.z
```
Comment (destogl) [inline README.md:345]: some interfaces are missing.

```suggestion
  - tcp_fts_sensor/force.x
  - tcp_fts_sensor/force.y
  - tcp_fts_sensor/force.z
  - tcp_fts_sensor/torque.x
  - tcp_fts_sensor/torque.y
  - tcp_fts_sensor/torque.z
```

File: README.md [hunk 37]
```diff
@@ -370,43 +370,42 @@ Notes:
...
 Available controllers:
-  - `joint_state_broadcaster[joint_state_broadcaster/JointStateBroadcaster]`
-  - `diffbot_base_controller[diff_drive_controller/DiffDriveController] active`
+- `forward_position_controller[forward_command_controller/ForwardCommandController]`
+- `fts_broadcaster[force_torque_sensor_broadcaster/ForceTorqueSensorBroadcaster]`
+- `joint_state_broadcaster[joint_state_broadcaster/JointStateBroadcaster]`
 
-Sending commands to diff drive controller:
+Notes:
+  - Wrench messages are not displayed properly in Rviz as NaN values are not handled in Rviz and FTS Broadcaster may send NaN values.
```
Comment (destogl) [inline README.md:394]:
```suggestion
```

**Review by bailaC — APPROVED** [review]
Rviz changes looks ok.

**Review by bmagyar — APPROVED** [review]

**Pull request merged**